include/proxy-wasm/context_interface.h

```diff
@@ -0,0 +1,622 @@
+// Copyright 2016-2019 Envoy Project Authors
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "include/proxy-wasm/compat.h"
+
+#include <time.h>
+#include <atomic>
+#include <chrono>
+#include <functional>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <vector>
+
+namespace proxy_wasm {
+
+#include "proxy_wasm_common.h"
+#include "proxy_wasm_enums.h"
+
+using Pairs = std::vector<std::pair<string_view, string_view>>;
+using PairsWithStringValues = std::vector<std::pair<string_view, std::string>>;
+using HttpCallToken = uint32_t;
+using GrpcToken = uint32_t;
+using SharedQueueDequeueToken = uint32_t;
+using SharedQueueEnqueueToken = uint32_t;
+
+// TODO: move to SDK.
+enum WasmStreamType {
+  Request,
+  Response,
+  Downstream,
+  Upstream,
+};
+
+// TODO: update SDK.
+enum class ProxyAction : uint32_t {
+  Illegal = 0,
+  Continue = 1,
+  Pause = 2,
+};
+
+struct PluginBase;
+class WasmBase;
+
+/**
+ * BufferInterface provides a interface between proxy-specific buffers and the proxy-independent ABI
+ * implementation. Embedders should subclass BufferInterface to enable the proxy-independent code to
+ * implement ABI calls which use buffers (e.g. the HTTP body).
+ */
+struct BufferInterface {
+  virtual ~BufferInterface() {}
+  virtual size_t size() const = 0;
+  /**
+   * Copy bytes from the buffer into the Wasm VM corresponding to 'wasm'.
+   * @param start is the first buffer byte to copy.
+   * @param length is the length of sequence of buffer bytes to copy.
+   * @param ptr_ptr is the location in the VM address space to place the address of the newly
+   * allocated memory block which contains the copied bytes.
+   * @param size_ptr is the location in the VM address space to place the size of the newly
+   * allocated memory block which contains the copied bytes (i.e. length).
```

Comment:
I think I asked this previously, but will this always end up equal to length? Or might it be less than length if the buffer was not length size for example? It would be good to clarify this a bit. At face value it's not clear to me why the length gets copied back out.

Reply:
This is just convenience.  It is copied out because otherwise we would need an additional call to copy it to size_ptr, and so there would always be a pair of calls.  This is more concise and safer because there is no way to forget to set the size or to get the size wrong.